pandas/core/groupby/indexing.py

```diff
@@ -0,0 +1,227 @@
+from __future__ import annotations
+
+from typing import (
+    Iterable,
+    cast,
+)
+
+import numpy as np
+
+from pandas._typing import (
+    FrameOrSeries,
+    PositionalIndexer,
+)
+from pandas.util._decorators import (
+    cache_readonly,
+    doc,
+)
+
+from pandas.core.dtypes.common import (
+    is_integer,
+    is_list_like,
+)
+
+from pandas.core.groupby import groupby
+from pandas.core.indexes.api import CategoricalIndex
+
+
+class GroupByIndexingMixin:
+    """
+    Mixin for adding .rows to GroupBy.
+    """
+
+    @cache_readonly
+    def _rows(self) -> _rowsGroupByIndexer:
+        return _rowsGroupByIndexer(cast(groupby.GroupBy, self))
+
+
+@doc(GroupByIndexingMixin._rows)
+class _rowsGroupByIndexer:
```

Comment:
its ok and preferred to use capitalcase, e.g. RowsGroupByIndexer)

Reply:
> its ok and preferred to use capitalcase, e.g. RowsGroupByIndexer

Done

> ok this is just for typing, so if you put it in a TYPE_CHECKING block will be ok (e.g .as the groupby.GroupBy should be separated out to avoid deps)

Unfortunately, I need groupby because I need `groupby.group_selection_context(self.groupby_object)`.

Reply:
@reback @jbrockmendel 

> ok this is just for typing, so if you put it in a TYPE_CHECKING block will be ok (e.g .as the groupby.GroupBy should be separated out to avoid deps)
Unfortunately, I need groupby because I need groupby.group_selection_context(self.groupby_object)

OK, I have tested making `group_selection_context` a method of the GroupBy class, rather than a module scoped method. As far as I can see everything works OK. But it is a bit drastic. My little project seems to be creeping in scope...

Unless you say otherwise, I will make that change. Of course, it effects the entire system, but I suppose we could always back it out...

Reply:
> @Reback @jbrockmendel
> 
> > ok this is just for typing, so if you put it in a TYPE_CHECKING block will be ok (e.g .as the groupby.GroupBy should be separated out to avoid deps)
> > Unfortunately, I need groupby because I need groupby.group_selection_context(self.groupby_object)
> 
> OK, I have tested making `group_selection_context` a method of the GroupBy class, rather than a module scoped method. As far as I can see everything works OK. But it is a bit drastic. My little project seems to be creeping in scope...
> 
> Unless you say otherwise, I will make that change. Of course, it effects the entire system, but I suppose we could always back it out...

I've committed it. Fingers crossed...

Reply:
no we don't care, this is an internal routine (its not on a groupby rather a grouper which is private)